### ISSUE TITLE
3.3.3 Error Suggestion - clarify the most common security exception

### ISSUE THREAD

**Issue by goodwitch** [w3c/wcag#2175]
Dear AGWG,

I've always believed that there is a security exception for when a user types an incorrect login or password.  I believe it is a known security risk to require a specific error suggestion that says, for example, "login is incorrect." or "password is incorrect."   ...because this creates a security risk (giving hackers the ability to brute force figure out valid logins and valid passwords).

But when I try to find that detail in WCAG 2.x, I can't find it!  I just have to magically know it.

I think it would be super helpful to include this exception example in the Understanding for 3.3.3 https://www.w3.org/TR/UNDERSTANDING-WCAG20/minimize-error-suggestions.html 

It would also be helpful to add the exception to the testing process for G83 https://www.w3.org/TR/2016/NOTE-WCAG20-TECHS-20161007/G83 

Resource:  a reference to why "login is incorrect"  or "password is incorrect" is a security risk - https://cheatsheetseries.owasp.org/cheatsheets/Authentication_Cheat_Sheet.html

**Comment by awkawk** [comment]
I don't think that 3.3.3 would apply since suggestions are not known. I think 3.3.1 is the challenge here.

**Comment by patrickhlauke** [comment]
agree this feels like 3.3.1 rather than 3.3.3.

**Comment by goodwitch** [comment]
Okay...but here is the rub, the clause "unless it would jeopardize the security or purpose of the content." is only in the normative of 3.3.3 (not even a hint of that exception in 3.3.1).  

> 3.3.3 Error Suggestion (Level AA): If an input error is automatically detected and suggestions for correction are known, then the suggestions are provided to the user, **unless it would jeopardize the security** or purpose of the content.

So, I've got members of my team trying to call a WCAG 2.x failure on username and password fields when the error message reads something like:

<ul>
  <li>If the username is CORRECT and the password is INCORRECT <br/>
  OR<br/>
  If the username is INCORRECT and the password is CORRECT
  <ul>
    <li>
      "Invalid username or password. Suggestion: Enter a valid username and password."
    </li>
  </ul>
  </li>

  <li>If the username is BLANK and the password is CORRECT <br/>
  OR<br/>
  If the username is CORRECT and the password is BLANK.
  <ul>
    <li>
      "Username and password are required. Suggestion: Enter a username and password."
    </li>
  </ul>
  </li>
</ul>

And what I need (and the world of a11y experts needs) is clarification on what is meant by "unless it would jeopardize the security" in the normative of 3.3.3.

**Comment by JAWS-test** [comment]
I think the exception in 3.3.3 refers to the fact that the application (logically) does not have to reveal the username and password, although it knows both.

What would actually be needed in 3.3.1 is a new exception (which does not exist so far) that the incorrect field does not have to be named if security reasons prohibit this

**Comment by awkawk** [comment]
I'm not sure that an exception is needed, but probably would be good to clarify in the Understanding document that there are cases where combinations of input fields are in error, so an error message could covers both inputs and in doing so can identify the error.

For example, "The username and password combination is not correct" when the username and/or password is not validating, or "The date you have selected is not valid" when the user selects "February" from a month dropdown and "30" from a day dropdown (and yes, the developer should modify the day drop down depending on the month, but didn't in this example).

**Comment by patrickhlauke** [comment]
> the username and password, although it knows both.

that would imply the password is stored as clear text, which...is certainly not an advisable thing

**Comment by awkawk** [comment]
@goodwitch I don't think that you are going to get crisp clarity on what counts as jeopardizing security in 3.3.3 from the WG. I think that like "equivalent", "security" has some level of subjectivity that site owners need to assess for themselves.

**Comment by goodwitch** [comment]
@awkawk I'm not looking for crisp clarity...but I sure would appreciate a little defuzzing.  There is zero write up about what was meant by "unless it would jeopardize the security or purpose of the content."

I'm simply suggesting that an example of each of these (security AND purpose) would benefit the world.  You know...in helping a11y testing be move objective with repeatable with similar results.

Can we add something like:

The clause "unless it would jeopardize the security or purpose of the content" does allow for valid exceptions.  A few examples of valid exceptions are:

<ul>
<li>Security Exception: When a user types an incorrect login or password it is a known security risk to require a specific error suggestion that says, for example, "login is incorrect." or "password is incorrect."  Why?  Because this creates a security risk making it possible for hackers to figure out valid logins and valid passwords.  
<ul>
<li>Acceptable solution:  Provide an error message like: "Invalid username or password. Suggestion: Enter a valid username and password."</li>
</ul>
<li>Purpose Exception:  If providing an explicit error message negates the purpose of the content commonly happens in learning situations.  For example, if the activity is an exam to test a person knowledge...and the point of a specific question is to select all the correct answers to a multiple choice question..and the user has only selected 2 of the 3 correct answers, providing an explicit error message telling the user they have only selected 2 of the 3 correct answers would negate the point of the question.
<ul>
<li>Acceptable solution:  No error message is required in this situation. Do not give away answers to an online exam. </li>
</ul>

**Comment by awkawk** [comment]
> Security Exception: When a user types an incorrect login or password it is a known security risk to require a specific error suggestion that says, for example, "login is incorrect." or "password is incorrect." Why? Because this creates a security risk making it possible for hackers to figure out valid logins and valid passwords.
> Acceptable solution: Provide an error message like: "Invalid username or password. Suggestion: Enter a valid username and password."

That sounds good to me. You should do a pull request on the understanding document... :)

> Purpose Exception: If providing an explicit error message negates the purpose of the content commonly happens in learning situations. For example, if the activity is an exam to test a person knowledge...and the point of a specific question is to select all the correct answers to a multiple choice question..and the user has only selected 2 of the 3 correct answers, providing an explicit error message telling the user they have only selected 2 of the 3 correct answers would negate the point of the question.
> Acceptable solution: No error message is required in this situation. Do not give away answers to an online exam.

Same for this!

These are both for 3.3.3, it sounded like you were having more trouble with 3.3.1 - no?

By the way, for your example earlier:

> If the username is BLANK and the password is CORRECT
> OR
> If the username is CORRECT and the password is BLANK.
> "Username and password are required. Suggestion: Enter a username and password."

I would say in this case that there is no way to know if the username or password are correct if the other is blank. While I'm fine with adding a "your username or password are not correct" message to the example before this, if a field is empty I think that it is reasonable to have a message that says "Username cannot be empty" or "Password cannot be blank".

**Comment by mraccess77** [comment]
I think this is a 3.3.1 exception as we don't want to identify the specific field that is in error -but  more generally would need to identify there was an error.

**Comment by bruce-usab** [comment]
@mraccess77 &mdash; are you saying that a normative  exception is needed to be added to 3.3.1?  IHMO 3.3.1/3.3.3 are okay as-is.
